### ISSUE TITLE
Most recent stable version is not selected when using `"*"` in queries

### ISSUE THREAD

**Issue by brendanzab** [balsoft/opam-nix#53]
**Describe the bug**
opam-nix selects `ocaml 5.0.0~rc1` over `ocaml 5.0.0` when using `ocaml-base-compiler = "*"`

**To Reproduce**

- Clone https://github.com/brendanzab/language-garden.git and checkout https://github.com/brendanzab/language-garden/commit/97f9f2d460a5894b4b2adf0a1612aec902b7d4e0
- Check the `ocaml` version:
  ```command
  $ nix develop --command ocaml --version
  The OCaml toplevel, version 5.0.0
  ```
- Apply the following diff:
  ```diff
  diff --git a/flake.nix b/flake.nix
  index 012f4aa..2134b1d 100644
  --- a/flake.nix
  +++ b/flake.nix
  @@ -72,7 +72,7 @@
         # when building the opam project with `opam-nix`.
         query = devPackagesQuery // {
           # Force the ocaml compiler to be taken from opam-repository:
  -        ocaml-base-compiler = "5.0.0"; # Fix for opam-nix choosing 5.0.0~rc1 over 5.0.0
  +        ocaml-base-compiler = "*";
         };
  
         # Package-specific derivation overrides.
  ```
- Check the `ocaml` version:
  ```command
  $ nix develop --command ocaml --version
  The OCaml toplevel, version 5.0.0~rc1
  ```

**Expected behavior**
opam-nix would select the most recent stable version when `"*"` is specified. Alternatively it would select the most recent _unstable_ version (there are more recent versions of [`ocaml-base-compiler`](https://github.com/ocaml/opam-repository/tree/8241abf97ddbcb63f614372f31abeb306312f33c/packages/ocaml-base-compiler) now).

**Environment**
 - OS name + version: macOS 13.4.1
 - Version of the code: e83bd1d949c5e330a49f89d394b51b744248e3ca

**Comment by balsoft** [comment]
`ocaml-base-compiler = "*"` just gets passed along to `opam` as `--packages ocaml-base-compiler`, which should select the latest version (to some definition of "latest"). I believe `5.0.0` is a release, and thus should be "stable"?

**Comment by brendanzab** [comment]
Ahh, wonder why it is not selecting it 😕 – the docs on version ordering are [here](https://opam.ocaml.org/doc/Manual.html#version-ordering). 

> The `~` character is special as it sorts even before the end of sequence (i.e. before anything shorter: "\~" sorts before "", "a\~b" before "a"). It's most convenient for pre-releases, allowing `1.0~beta` to be before `1.0`.

**Comment by balsoft** [comment]
Not sure what else you are expecting? It seems to be behaving according to the documentation.

**Comment by balsoft** [comment]
You might to be looking at the wrong version of opam-repository; mind that by default `opam-nix` uses the version pinned in flake.lock (at this point, it's 786c55fa77c37f07eceea7d6a9bec04d2225e302)

**Comment by brendanzab** [comment]
> Not sure what else you are expecting? It seems to be behaving according to the documentation.

I would have expected it to select the most recent non-prerelease version (i.e. nothing after the tilde), based on what is on the pinned version of `opam-repository`. But perhaps I don’t understand opam versioning that well (I’m used to how it works with Cargo).

Maybe I should just set the compiler version explicitly to avoid this in the future, as opposed to relying on `*`?

As an aside, is there a good way to query what versions of a package in the local `opam-repository`? At the moment I have just been going into the github repository to check what is at the current commit pinned in my flake, but that seems a little round-about. I assume there’s probably some OCaml tool that can help?

> You might to be looking at the wrong version of opam-repository; mind that by default opam-nix uses the version pinned in flake.lock (at this point, it's 786c55fa77c37f07eceea7d6a9bec04d2225e302)

I was looking at the commit of `opam-repository` that I pinned on my flake. Sorry, I probably should have mentioned that – I was planning to, but I must have forgot. In the [commit I linked to](https://github.com/brendanzab/language-garden/commit/97f9f2d460a5894b4b2adf0a1612aec902b7d4e0) it was https://github.com/ocaml/opam-repository/commit/8241abf97ddbcb63f614372f31abeb306312f33c. I guess I’m wondering if this is exposing an issue that is not obvious in the currently pinned version of `opam-repository`?

**Comment by greedy** [comment]
I tracked this down to the criteria used for dependency resolution in opam-nix. The current default criteria is `-count[version-lag,request],-count[version-lag,changed],-count[avoid-version,request]`. Specifically, the position of the `-count[avoid-version,request]` leads to this for my situation; if it is first then the prerelease compiler versions (which have the avoid-version) flag are no longer selected. I think the difference putting it first is that avoiding avoid-version packages is given higher weight than the other criteria which are somehow favoring the pre-release compilers; I haven't bothered figuring out exactly what the version-lag criteria means.

TL;DR changing the resolution criteria to opam's current default of `-removed,-count[avoid-version,changed],-count[version-lag,request],-count[version-lag,changed],-count[missing-depexts,changed],-changed` is solving this problem for me.

Also, can just set `criteria = null` in resolveArgs to automatically get opam's default. Is the reason that criteria is specified in defaultResolveArgs just to improve the stability of resolution as opam is updated?

**Comment by balsoft** [comment]
> Also, can just set criteria = null in resolveArgs to automatically get opam's default. Is the reason that criteria is specified in defaultResolveArgs just to improve the stability of resolution as opam is updated?

Yes, this is the original reason. Although looking at this, maybe it's not the best idea, as the current criteria used to be opam's default at some point, and now they're clearly broken.

Thanks for figuring this out!

**Comment by balsoft** [comment]
Is this fixed by #110?

**Comment by balsoft** [comment]
Closing under the assumption that this is fixed
